### DISCUSSION TITLE
Thread-safe memory object streams

### DISCUSSION THREAD

**Discussion by davidbrochart** [agronholm/anyio#521] [Q&A]
[Memory object streams](https://anyio.readthedocs.io/en/stable/streams.html#memory-object-streams) are not thread-safe. Since they don't (de)serialize, they seem to be a good fit to pass data between different threads.
In a scenario where there is an event loop running in the main thread, and a separate event loop running in another thread, how can memory object streams be used to pass data between both threads?
This hangs:
```py
from functools import partial
from anyio import create_memory_object_stream, create_task_group, run, sleep, to_thread

async def process_items(send_stream):
    async with send_stream:
        for num in range(10):
            await sleep(1)
            await send_stream.send(f"number {num}")

async def main():
    send_stream, receive_stream = create_memory_object_stream()
    async with create_task_group() as tg:
        # tg.start_soon(process_items, send_stream)  # works fine
        tg.start_soon(to_thread.run_sync, partial(run, partial(process_items, send_stream)))
        async with receive_stream:
            async for item in receive_stream:
                print("received", item)

run(main)
```

**Comment by agronholm** [comment]
In order to send messages between event loop threads, you would have to create a memory object stream and start a blocking portal in one event loop, and then use `to_thread.run_sync()` in the other one to use that portal to call `send()` on that memory object stream.

**Reply by agronholm** [reply]
May I ask why you have multiple event loop threads running?

**Reply by davidbrochart** [reply]
I have the main thread running an event loop, but it can occasionally run blocking (non-async) code in its event loop. I have a separate thread that must run async code in parallel, and it must not be blocked by the main thread. So I'm not in [this situation](https://anyio.readthedocs.io/en/stable/threads.html#calling-asynchronous-code-from-a-worker-thread), where I launch my thread with `to_thread` and run async code with `from_thread`, right? Because the thread could be blocked by the main thread.

**Reply by agronholm** [reply]
So, if I'm understanding correctly, you have something that mainly runs blocking code but occasionally needs to call async code?
Is this second thread spawned by a third party library, out of your control?

**Reply by davidbrochart** [reply]
Let's say that the main thread can run blocking code in its event loop (it's not known how much).
The second thread could be spawned externally, but if that's easier I could spawn it. Do you see a simpler solution if I spawn it?

**Reply by agronholm** [reply]
if you spawn it, then you could just naturally call async code from that thread, without having to run a parallel event loop.

**Reply by davidbrochart** [reply]
But then that thread could be blocked by the main thread (which can run blocking code), and I don't want that because other threads depend on that thread.

**Reply by agronholm** [reply]
Now you lost me. How could one thread block another by running blocking code?

**Reply by davidbrochart** [reply]
If the main thread's event loop is blocked (because running blocking code), then the other thread running async code in the main thread's event loop is also blocked, right?

**Reply by agronholm** [reply]
Yeah, when running async code, this is true. But couldn't you run that blocking code in worker threads? That's how it's supposed to be used. How long do you expect the blocking code to block the event loop, per call?

**Reply by davidbrochart** [reply]
No, I cannot run that blocking code in another thread, it has to be run in the main thread.
Maybe to give some context, I should say that I'm planning to rewrite [ipykernel](https://github.com/ipython/ipykernel) using `AnyIO`. In Jupyter, the (blocking) code comes from the user, and we cannot know in advance what it's going to be. Also, the user expects to run their code in the main thread (some libraries behave differently otherwise).

**Reply by agronholm** [reply]
Ok, so if I understand you correctly, then my original answer is appropriate: the combination of a blocking portal and `to_thread.run_sync()`.

**Reply by davidbrochart** [reply]
Thanks, I have to admit that I don't quite see how to implement it. Could you please show a minimal example?

**Reply by davidbrochart** [reply]
In particular, `to_thread.run_sync()` expects a regular function, so I cannot call the memory object stream's `send()`, unless I wrap it in a regular function. In the end, that's what I have and it still hangs:
```py
from anyio import start_blocking_portal, create_memory_object_stream, create_task_group, run, sleep, to_thread, from_thread

def send_wrapper(send_stream, num):
    from_thread.run(send_stream.send, f"number {num}")

async def process_items(send_stream):
    async with send_stream:
        for num in range(10):
            await sleep(1)
            await to_thread.run_sync(send_wrapper, send_stream, num)

async def main(portal):
    send_stream, receive_stream = create_memory_object_stream()
    portal.start_task_soon(process_items, send_stream)
    async with create_task_group() as tg:
        async with receive_stream:
            async for item in receive_stream:
                print("received", item)

with start_blocking_portal() as portal:
    run(main, portal)
```

**Reply by davidbrochart** [reply]
This doesn't use a blocking portal and seems to work:
```py
from anyio import create_memory_object_stream, create_task_group, run, sleep, to_thread, from_thread

def process_items(send_stream, tg):
    async def main():
        async with send_stream:
            for num in range(10):
                await sleep(1)
                item = f"number {num}"
                from_thread.run_sync(tg.start_soon, send_stream.send, item)
                print("sent", item)
    run(main)

async def main():
    send_stream, receive_stream = create_memory_object_stream()
    async with create_task_group() as tg:
        tg.start_soon(to_thread.run_sync, process_items, send_stream, tg)
        async with receive_stream:
            async for item in receive_stream:
                print("received", item)

run(main)
```

**Reply by davidbrochart** [reply]
Ah but if the main thread blocks then the other thread blocks too... I'm kind of lost now.

**Reply by davidbrochart** [reply]
I think the solution would be something like https://github.com/aio-libs/janus.

**Reply by davidbrochart** [reply]
I found a solution, but using sockets and thread-safe queues (to do something similar to `janus`), instead of AnyIO's streams:
```py
import queue
import time

import zmq.asyncio
import zmq
from anyio import create_task_group, run, sleep, to_thread


def process_items(q, socket_send):
    async def main():
        for num in range(10):
            await sleep(1)
            item = f"number {num}"
            q.put(item)
            await socket_send.send(b"")  # notify receiver
            print("sent", item)
    run(main)

async def main():
    q = queue.Queue()
    context = zmq.asyncio.Context()
    socket_recv = context.socket(zmq.PAIR)
    socket_recv.bind("inproc://event")
    socket_send = context.socket(zmq.PAIR)
    socket_send.connect("inproc://event")
    async with create_task_group() as tg:
        tg.start_soon(to_thread.run_sync, process_items, q, socket_send)
        while True:
            await socket_recv.recv()
            item = q.get()  # we know there is an item
            print("received", item)
            time.sleep(2)  # simulate blocking code

run(main)
```
